### PR TITLE
[hardware, fsdp, vllm, trainer, ray] feat: add Intel XPU support for e2e GRPO/PPO/SFT training

### PR DESCRIPTION
**Pull request by kahlun**
# What does this PR do?
End-to-end Intel XPU (Arc Pro / Data Center GPU Max) support for GRPO, PPO, and SFT training using FSDP + vLLM rollout.

This is the follow-up to #5943 (device detection) and bundles all remaining XPU enablement work into a working e2e stack. Follows the same incremental pattern as the Ascend NPU enablement. Validated on 2× Intel Arc Pro B70 (32GB, Battlemage, driver 26.09).

## Checklist Before Starting
 PR title: [hardware, fsdp, vllm, trainer, ray] feat: add Intel XPU support for e2e GRPO/PPO/SFT training
## Test
Hardware: 2× Intel Arc Pro B70 (32 GB PCIe, Battlemage), PyTorch 2.11+xpu, driver 26.09
Hardware 4x Intel Arc Pro B60 (24 GB PCIe, Battlemage), Virtual Machine

## GRPO smoke test (tests/special_xpu/run_grpo_xpu.sh) — Qwen2.5-0.5B-Instruct, GSM8K, 1 step:
Training Progress: 100%|██████████| 1/1 [00:51<00:00, 51.68s/it]
step:1 - actor/entropy:4.64 - actor/loss:0.00482 - actor/grad_norm:1.45
       - actor/lr:5e-07 - timing_s/step:51.59 - perf/throughput:73.2
Validation (step:0) and training (step:1) complete without crashes or NaN loss. SFT and PPO scripts also included.

## API and Usage Example
No API changes. XPU is auto-detected via torch.xpu.is_available(); existing CUDA/NPU configs unchanged.

# Build
docker build -t verl-xpu:latest -f docker/xpu/Dockerfile.xpu .

# Run GRPO (2 GPUs)
docker run --rm --privileged --device /dev/dri --ipc=host --shm-size 10g \
  -v $HOME/data:/root/data -e NUM_GPUS=2 verl-xpu:latest \
  bash -c "source /root/.bashrc && cd /workspace/verl && bash tests/special_xpu/run_grpo_xpu.sh"

## Design & Code Changes
25 files, +849/-32. Consolidates #5943 (device detection) and all remaining XPU enablement work into a single self-contained PR. 

#Key areas:
## Infrastructure

docker/xpu/Dockerfile.xpu — image based on intel/deep-learning-essentials:2025.3.2, pinned driver 26.09
requirements-xpu.txt, tests/special_xpu/ (GRPO/PPO/SFT smoke scripts)
Ray integration (single_controller/ray/base.py, worker.py)

## Ray's IntelGPUAccelerator maps XPU → "GPU" slot; XPU workers treated identically to CUDA
Sync ONEAPI_DEVICE_SELECTOR from ZE_AFFINITY_MASK on worker start so Level Zero re-numbers masked devices correctly
Distributed (utils/distributed.py, fsdp_utils.py)

## oneccl fixes
The upstream from pytorch xpu fixes timeline (pytorch2.13 +)
all_reduce_avg(): xccl ReduceOp.AVG double-divides on small tensors ([[torch-xpu-ops#3020](https://github.com/intel/torch-xpu-ops/issues/3020)]); uses SUM÷N instead
Composite backend cpu:gloo,xpu:xccl for mixed CPU/XPU process groups
set_force_sum_reduction_for_comms(True) for FSDP2

##VLLM
vLLM coexistence (vllm_async_server.py, xpu_patches.py)
Force enable_sleep_mode=False (XPU MemPool unavailable) 
Pop ONEAPI_DEVICE_SELECTOR before AsyncLLM init — vLLM spawns subprocesses where ONEAPI_DEVICE_SELECTOR=level_zero:* triggers SIGABRT in SYCL JIT; ZE_AFFINITY_MASK alone is sufficient
distributed_executor_backend="uni" for TP=1

Idempotent monkey-patches for false OOM check ([https://github.com/vllm-project/vllm/pull/37149], draft XpuMemAllocator) and profiling assert ([https://github.com/vllm-project/vllm/pull/36720], same assert on ROCm; XPU fix not yet upstream) — version-guarded, self-remove once upstream lands
Other

## some conditional changes or enhancement
model.py: read attn_implementation from config instead of hardcoding flash_attention_2
replica.py: replace hardcoded "cuda"/"npu" with get_device_name()
bucketed_weight_transfer.py: guard ipc_collect() with hasattr
transformer_impl.py: register "xpu" in @EngineRegistry
constants_ppo.py: propagate ONEAPI_DEVICE_SELECTOR into Ray workers' runtime_env

## Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] pre-commit checks passed: `pre-commit run --files verl/utils/device.py` — all 12 checks passed (ruff, mypy, license, etc.)
- [x] Documentation: docker/xpu/README.md included
- [x] Unit tests: XPU detection requires XPU hardware; validated manually on Intel Arc Pro B70. CI cannot run on XPU hardware.
- [ ] CI request: no XPU runners available (same situation as NPU); will post in #ci-request once ready
- [x] Not related to recipe submodule.

  
## Scope: 
This PR covers FSDP + vLLM rollout only (the primary verl training stack). TorchTitan (E1/E2) and VeOmni are excluded — TorchTitan XPU is functional on Battlemage but not yet validated on all hardware; those can follow as separate PRs. flash_attention_2 is replaced with pure-PyTorch SDPA throughout XPU paths. IPC (is_support_ipc=False) is disabled for XPU because the CUDA 8-handle IPC protocol is incompatible with SYCL handles ([https://github.com/intel/torch-xpu-ops/issues/1678]); the shared-memory fallback is used instead.

### CODE REVIEW

**Comment by kahlun** [comment]
"Note: this is a draft PR submitted early by mistake. Still being cleaned up — will mark ready for review once docs and pre-commit checks are done. Sorry for the noise."

Still checking on the PR message. Sorry for any 'requested review' appeared.

**Comment by wuxibin89** [comment]
https://github.com/verl-project/verl/pull/6119

**Comment by kahlun** [comment]
Withdrawing this contribution for internal reasons.
